src/lib/device_tree/tests/integration_test.rs

```diff
@@ -71,3 +76,78 @@ fn test_reads_properties() {
     assert_property(it.next().unwrap(), "#address-cells", 2, &vec![0, 0, 0, 1]);
     assert!(it.next().is_none());
 }
+
+#[test]
+fn test_reads_empty_properties() {
+    let data = dts_to_dtb(
+        r#"
+/dts-v1/;
+/ { 
+    #address-cells;
+};"#,
+    );
+    let slice_reader = &SliceReader::new(&data);
+    let reader = FdtReader::new(slice_reader).unwrap();
+    let mut it = reader.walk();
+
+    assert_node(it.next().unwrap(), "", 1);
+    assert_property(it.next().unwrap(), "#address-cells", 2, &vec![]);
+    assert!(it.next().is_none());
+}
+
+#[test]
+fn test_reads_nested_nodes() {
+    let data = dts_to_dtb(
+        r#"
+/dts-v1/;
+/ {
+    node1 {
+        #address-cells = "ok";
+        node2 {
+        };
+    };
+    node3 {
+    };
+};"#,
+    );
+    let slice_reader = &SliceReader::new(&data);
+    let reader = FdtReader::new(slice_reader).unwrap();
+    let mut it = reader.walk();
+
+    assert_node(it.next().unwrap(), "", 1);
+    assert_node(it.next().unwrap(), "node1", 2);
+    assert_property(it.next().unwrap(), "#address-cells", 3, &vec![111, 107, 0]);
+    assert_node(it.next().unwrap(), "node2", 3);
+    assert_node(it.next().unwrap(), "node3", 2);
+    assert!(it.next().is_none());
+}
+
+#[test]
+#[should_panic(expected = "invalid magic in device tree header")]
+fn test_returns_error_when_magic_is_invalid() {
+    let mut data = dts_to_dtb(
+        r#"
+/dts-v1/;
+/ {
+    #address-cells = "ok";
+};"#,
+    );
+    // Magic is a first number
+    data[0] = 1;
+    let slice_reader = &SliceReader::new(&data);
+    FdtReader::new(slice_reader).unwrap();
+}
+
+#[test]
+#[should_panic(expected = "not enough data to read device tree header")]
```

Comment:
Probably out of scope for this PR, but I find it very un-Rusty to panic instead of returning a `Result::Err`. I would imagine that there are contexts in which this could be recoverable.

Reply:
I converted `should_panic` tests into

```rust
assert_eq!(func().err, Some("...."))
```

To verify that function returns an expected error on invalid input. I think it is also cleaner now that we expect error in specific place, instead of catching any panic in this function.